server/src/instant/db/transaction.clj

```diff
@@ -141,36 +146,119 @@
+
+
+(defn resolve-lookups-for-delete-entity [conn app-id tx-steps]
+  (let [[delete-entity-steps tx-steps'] (coll/split-by #(= :delete-entity (first %)) tx-steps)
+        lookup-refs                     (->> delete-entity-steps
+                                             (map second)
+                                             (filter triple-model/eid-lookup-ref?))
+        resolved                        (resolve-lookups conn app-id lookup-refs)
+        delete-entity-steps'            (for [[op eid etype] delete-entity-steps]
+                                          [op (get resolved eid eid) etype])]
+    (concat delete-entity-steps' tx-steps')))
+
+(defn resolve-etypes-for-delete-entity [conn app-id tx-steps]
+  (let [[delete-entity-steps tx-steps'] (coll/split-by #(= :delete-entity (first %)) tx-steps)
+        untyped-ids                     (->> delete-entity-steps
+                                             (keep (fn [[_ id etype]]
+                                                     (when (nil? etype)
+                                                       id))))
+        resolved                        (resolve-etypes conn app-id untyped-ids)
+        delete-entity-steps'            (for [[op eid etype] delete-entity-steps
+                                              etype'         (if etype
+                                                               [etype]
+                                                               (get resolved eid))]
+                                          [op eid etype'])]
+    (concat delete-entity-steps' tx-steps')))
+
+(defn expand-delete-entity-cascade [conn app-id tx-steps]
+  (let [[delete-entity-steps tx-steps'] (coll/split-by #(= :delete-entity (first %)) tx-steps)]
+    (if (empty? delete-entity-steps)
+      tx-steps
+      (let [ids+etypes           (map next delete-entity-steps)
+            query                (str "WITH RECURSIVE entids (entity_id, etype) AS (
+                                       -- start with entid
+                                         VALUES " (string/join ", " (repeat (count ids+etypes) "(?, ?)")) "
+
+                                       -- recurse all 'cascade' refs
+                                       UNION
+                                         SELECT triples.entity_id , forward_ident.etype
+                                         FROM   triples
+                                         JOIN   attrs                   ON triples.attr_id = attrs.id
+                                         JOIN   idents AS forward_ident ON forward_ident.id = attrs.forward_ident
+                                         JOIN   idents AS reverse_ident ON reverse_ident.id = attrs.reverse_ident
+                                         JOIN   entids                  ON to_jsonb(entids.entity_id) = triples.value
+                                         WHERE  triples.app_id = ?
+                                         AND    attrs.on_delete = 'cascade'
```

Comment:
I wonder if you could avoid the join on attrs in the database by using the `attrs` object that we pass in to transact!